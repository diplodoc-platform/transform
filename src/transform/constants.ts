import type {Lang} from './typings';

export const DEFAULT_LANG: Lang = 'ru';
