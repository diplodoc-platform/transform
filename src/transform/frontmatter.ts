import {YAMLException, dump, load} from 'js-yaml';
import {dedent} from 'ts-dedent';
import cloneDeepWith from 'lodash/cloneDeepWith';

import {log} from './log';

export type FrontMatter = {
    [key: string]: unknown;
    metadata?: Record<string, unknown>[];
};

const SEP = '---';

/**
 * Temporary workaround to enable parsing YAML metadata from potentially
 * Liquid-aware source files
 * @param content Input string which could contain Liquid-style substitution syntax (which clashes with YAML
 * object syntax)
 * @returns String with `{}` escaped, ready to be parsed with `js-yaml`
 */
const escapeLiquid = (content: string): string =>
    content.replace(/{{/g, '(({{').replace(/}}/g, '}}))');

/**
 * Inverse of a workaround defined above.
 * @see `escapeLiquidSubstitutionSyntax`
 * @param escapedContent Input string with `{}` escaped with backslashes
 * @returns Unescaped string
 */
const unescapeLiquid = (escapedContent: string): string =>
    escapedContent.replace(/\(\({{/g, '{{').replace(/}}\)\)/g, '}}');

const matchMetadata = (fileContent: string) => {
    if (!fileContent.startsWith(SEP)) {
        return null;
    }

    const closeStart = fileContent.indexOf('\n' + SEP, SEP.length);
    const closeEnd = fileContent.indexOf('\n', closeStart + 1);

    if (closeStart === -1) {
        return null;
    }

    return [fileContent.slice(SEP.length, closeStart).trim(), fileContent.slice(closeEnd + 1)];
};

const duplicateKeysCompatibleLoad = (yaml: string, filePath: string | undefined) => {
    try {
        return load(yaml);
    } catch (e) {
        if (e instanceof YAMLException) {
            const duplicateKeysDeprecationWarning = dedent`
                In ${filePath ?? '(unknown)'}: Encountered a YAML parsing exception when processing file metadata: ${e.reason}.
                It's highly possible the input file contains duplicate mapping keys.
                Will retry processing with necessary compatibility flags.
                Please note that this behaviour is DEPRECATED and WILL be removed in a future version
                without further notice, so the build WILL fail when supplied with YAML-incompatible meta.
            `;

            log.warn(duplicateKeysDeprecationWarning);

            return load(yaml, {json: true});
        }

        throw e;
    }
};

export const extractFrontMatter = (
    fileContent: string,
    filePath?: string,
): [FrontMatter, string] => {
    const matches = matchMetadata(fileContent);

    if (matches) {
        const [metadata, strippedContent] = matches;

        return [
            cloneDeepWith(
                duplicateKeysCompatibleLoad(escapeLiquid(metadata), filePath) as FrontMatter,
                (v) => (typeof v === 'string' ? unescapeLiquid(v) : undefined),
            ),
            strippedContent,
        ];
    }

    return [{}, fileContent];
};

export const composeFrontMatter = (frontMatter: FrontMatter, strippedContent: string) => {
    const dumped = dump(frontMatter, {lineWidth: -1}).trim();

    // This empty object check is a bit naive
    // The other option would be to check if all own fields are `undefined`,
    // since we exploit passing in `undefined` to remove a field quite a bit
    if (dumped === '{}') {
        return strippedContent;
    }

    return `${SEP}\n${dumped}\n${SEP}\n${strippedContent}`;
};
