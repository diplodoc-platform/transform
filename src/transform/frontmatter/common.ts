export type FrontMatter = {
    [key: string]: unknown;
    metadata?: Record<string, unknown>[];
};

export const frontMatterFence = '---';

/**
 * Temporary workaround to enable parsing YAML metadata from potentially
 * Liquid-aware source files
 * @param content Input string which could contain Liquid-style substitution syntax (which clashes with YAML
 * object syntax)
 * @returns String with `{}` escaped, ready to be parsed with `js-yaml`
 */
export const escapeLiquidSubstitutionSyntax = (content: string): string =>
    content.replace(/{{/g, '(({{').replace(/}}/g, '}}))');

/**
 * Inverse of a workaround defined above.
 * @see `escapeLiquidSubstitutionSyntax`
 * @param escapedContent Input string with `{}` escaped with backslashes
 * @returns Unescaped string
 */
export const unescapeLiquidSubstitutionSyntax = (escapedContent: string): string =>
    escapedContent.replace(/\(\({{/g, '{{').replace(/}}\)\)/g, '}}');

export const countLineAmount = (str: string) => str.split(/\r?\n/).length;
