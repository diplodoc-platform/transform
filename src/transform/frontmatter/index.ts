export * from './extract';
export * from './emplace';
export * from './transformValues';
export {countLineAmount} from './common';
