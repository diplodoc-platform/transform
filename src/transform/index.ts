import {bold} from 'chalk';
import attrs from 'markdown-it-attrs';
import Token from 'markdown-it/lib/token';

import {log, LogLevels} from './log';
import makeHighlight from './highlight';
import extractTitle from './title';
import getHeadings from './headings';
import liquid from './liquid';
import sanitizeHtml, {SanitizeOptions} from './sanitize';

import notes from './plugins/notes';
import anchors from './plugins/anchors';
import code from './plugins/code';
import cut from './plugins/cut';
import deflist from './plugins/deflist';
import term from './plugins/term';
import file from './plugins/file';
import imsize from './plugins/imsize';
import meta from './plugins/meta';
import sup from './plugins/sup';
import tabs from './plugins/tabs';
import video from './plugins/video';
import monospace from './plugins/monospace';
import yfmTable from './plugins/table';
import {initMd} from './md';
import {MarkdownItPluginCb} from './plugins/typings';
import type {HighlightLangMap, Heading} from './typings';

interface OutputType {
    result: {
        html: string;
        title?: string;
        headings: Heading[];
        assets?: unknown[];
        meta?: object;
    };
    logs: Record<LogLevels, string[]>;
}
interface OptionsType {
    vars?: Record<string, string>;
    path?: string;
    extractTitle?: boolean;
    needTitle?: boolean;
    allowHTML?: boolean;
    linkify?: boolean;
    breaks?: boolean;
    conditionsInCode?: boolean;
    disableLiquid?: boolean;
    leftDelimiter?: string;
    rightDelimiter?: string;
    isLiquided?: boolean;
    needToSanitizeHtml?: boolean;
    sanitizeOptions?: SanitizeOptions;
    needFlatListHeadings?: boolean;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    plugins?: MarkdownItPluginCb<any>[];
    highlightLangs?: HighlightLangMap;
    root?: string;
    [x: string]: unknown;
}

function transform(originInput: string, opts: OptionsType = {}): OutputType {
    const {
        vars = {},
        path,
        extractTitle: extractTitleOption,
        needTitle,
        allowHTML = false,
        linkify = false,
        breaks = true,
        conditionsInCode = false,
        needToSanitizeHtml = false,
        sanitizeOptions,
        needFlatListHeadings = false,
        disableLiquid = false,
        leftDelimiter = '{',
        rightDelimiter = '}',
        isLiquided = false,
        plugins = [
            meta,
            deflist,
            cut,
            notes,
            anchors,
            tabs,
            code,
            sup,
            video,
            monospace,
            yfmTable,
            file,
            imsize,
            term,
        ],
        highlightLangs = {},
        ...customOptions
    } = opts;

    const pluginOptions = {
        ...customOptions,
        vars,
        path,
        extractTitle: extractTitleOption,
        disableLiquid,
        log,
    };

    const input =
        disableLiquid || isLiquided
            ? originInput
            : liquid(originInput, vars, path, {conditionsInCode});

    const highlight = makeHighlight(highlightLangs);
    const md = initMd({html: allowHTML, linkify, highlight, breaks});
    // Need for ids of headers
    md.use(attrs, {leftDelimiter, rightDelimiter});

    plugins.forEach((plugin) => md.use(plugin, pluginOptions));

    try {
        let title;
        let tokens;
        let titleTokens;
        const env = {} as {[key: string]: Token[] | unknown};

        tokens = md.parse(input, env);

        if (extractTitleOption) {
            ({title, tokens, titleTokens} = extractTitle(tokens));

            // title tokens include other tokens that need to be transformed
            if (titleTokens.length > 1) {
                title = md.renderer.render(titleTokens, md.options, env);
            }
        }
        if (needTitle) {
            ({title} = extractTitle(tokens));
        }

        const headings = getHeadings(tokens, needFlatListHeadings);

        // add all term template tokens to the end of the html
        const termTokens = (env.termTokens as Token[]) || [];
        let html = md.renderer.render([...tokens, ...termTokens], md.options, env);
        if (needToSanitizeHtml) {
            html = sanitizeHtml(html, sanitizeOptions);
        }

        const assets = md.assets;
        const meta = md.meta;

        return {
            result: {html, title, headings, assets, meta},
            logs: log.get(),
        };
    } catch (err) {
        log.error(`Error occurred${path ? ` in ${bold(path)}` : ''}`);
        throw err;
    }
}

export = transform;

// eslint-disable-next-line @typescript-eslint/no-namespace -- backward compatibility
namespace transform {
    export type Options = OptionsType;
    export type Output = OutputType;
}
