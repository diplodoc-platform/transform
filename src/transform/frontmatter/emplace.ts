import {dump} from 'js-yaml';

import {FrontMatter, frontMatterFence} from './common';

export const serializeFrontMatter = (frontMatter: FrontMatter) => {
    const dumped = dump(frontMatter, {lineWidth: -1}).trim();

    // This empty object check is a bit naive
    // The other option would be to check if all own fields are `undefined`,
    // since we exploit passing in `undefined` to remove a field quite a bit
    if (dumped === '{}') {
        return '';
    }

    return `${frontMatterFence}\n${dumped}\n${frontMatterFence}\n`;
};

export const emplaceSerializedFrontMatter = (
    frontMatterStrippedContent: string,
    frontMatter: string,
) => `${frontMatter}${frontMatterStrippedContent}`;

export const emplaceFrontMatter = (frontMatterStrippedContent: string, frontMatter: FrontMatter) =>
    emplaceSerializedFrontMatter(frontMatterStrippedContent, serializeFrontMatter(frontMatter));
