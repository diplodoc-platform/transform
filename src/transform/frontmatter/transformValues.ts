import {FrontMatter} from './common';

export const transformFrontMatterValues = (
    frontMatter: FrontMatter,
    valueMapper: (v: unknown) => unknown,
): FrontMatter => {
    const transformInner = (something: unknown): unknown => {
        if (Array.isArray(something)) {
            return something.map((el) => transformInner(el));
        }

        if (typeof something === 'object' && something !== null) {
            return Object.fromEntries(
                Object.entries(something).map(([k, v]) => [k, transformInner(v)]),
            );
        }

        if (typeof something === 'function') {
            return something;
        }

        return valueMapper(something);
    };

    return transformInner(frontMatter) as FrontMatter;
};
