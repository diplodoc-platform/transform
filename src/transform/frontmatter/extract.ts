import {YAMLException, load} from 'js-yaml';

import {log} from '../log';

import {
    FrontMatter,
    countLineAmount,
    escapeLiquidSubstitutionSyntax,
    frontMatterFence,
    unescapeLiquidSubstitutionSyntax,
} from './common';
import {transformFrontMatterValues} from './transformValues';

type ParseExistingMetadataReturn = {
    frontMatter: FrontMatter;
    frontMatterStrippedContent: string;
    frontMatterLineCount: number;
};

const matchMetadata = (fileContent: string) => {
    if (!fileContent.startsWith(frontMatterFence)) {
        return null;
    }

    // Search by format:
    // ---
    // metaName1: metaValue1
    // metaName2: meta value2
    // incorrectMetadata
    // ---
    const regexpMetadata = '(?<=-{3}\\r?\\n)((.*\\r?\\n)*?)(?=-{3}\\r?\\n)';
    // Search by format:
    // ---
    // main content 123
    const regexpFileContent = '-{3}\\r?\\n((.*[\r?\n]*)*)';

    const regexpParseFileContent = new RegExp(`${regexpMetadata}${regexpFileContent}`, 'gm');

    return regexpParseFileContent.exec(fileContent);
};

const duplicateKeysCompatibleLoad = (yaml: string, filePath: string | undefined) => {
    try {
        return load(yaml);
    } catch (e) {
        if (e instanceof YAMLException) {
            const duplicateKeysDeprecationWarning = `
                In ${filePath ?? '(unknown)'}: Encountered a YAML parsing exception when processing file metadata: ${e.reason}.
                It's highly possible the input file contains duplicate mapping keys.
                Will retry processing with necessary compatibility flags.
                Please note that this behaviour is DEPRECATED and WILL be removed in a future version
                without further notice, so the build WILL fail when supplied with YAML-incompatible meta.
            `
                .replace(/^\s+/gm, '')
                .replace(/\n/g, ' ')
                .trim();

            log.warn(duplicateKeysDeprecationWarning);

            return load(yaml, {json: true});
        }

        throw e;
    }
};

export const separateAndExtractFrontMatter = (
    fileContent: string,
    filePath?: string,
): ParseExistingMetadataReturn => {
    const matches = matchMetadata(fileContent);

    if (matches && matches.length > 0) {
        const [, metadata, , metadataStrippedContent] = matches;

        return {
            frontMatter: transformFrontMatterValues(
                duplicateKeysCompatibleLoad(
                    escapeLiquidSubstitutionSyntax(metadata),
                    filePath,
                ) as FrontMatter,
                (v) => (typeof v === 'string' ? unescapeLiquidSubstitutionSyntax(v) : v),
            ),
            frontMatterStrippedContent: metadataStrippedContent,
            frontMatterLineCount: countLineAmount(metadata),
        };
    }

    return {
        frontMatter: {},
        frontMatterStrippedContent: fileContent,
        frontMatterLineCount: 0,
    };
};
