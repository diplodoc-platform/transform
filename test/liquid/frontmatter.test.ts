import dedent from 'ts-dedent';

import {composeFrontMatter, extractFrontMatter} from '../../src/transform/frontmatter';
import liquidDocument from '../../src/transform/liquid';

describe('front matter extract/emplace utility function pair', () => {
    it.each([
        dedent`
        ---
        prop: value
        ---
        # Content.
        `,
        dedent`
        ---
        prop: value
        ---

        # Content.
        `,
        dedent`
        ---
        prop: value
        ---





        # Content.
        `,
        dedent`
        # Content.
        `,
        dedent`




        # Content.
        `,
    ])(
        `preserves the same amount of linebreaks between front matter block and content %#`,
        (input) => {
            const [frontMatter, strippedContent] = extractFrontMatter(input);
            const emplaced = composeFrontMatter(frontMatter, strippedContent);

            expect(emplaced).toEqual(input);
        },
    );

    it('is able to handle YAML with unquoted substitution syntax', () => {
        const content = dedent`
        ---
        prop: {{ wouldbreak }}
        note: This snippet typically shouldn't be able to get parsed, since {} is object syntax in YAML.
        ---

        Test.
        `;

        const [frontMatter] = extractFrontMatter(content);

        expect(frontMatter).toMatchObject({prop: '{{ wouldbreak }}'});
    });
});

describe('Liquid substitutions in front matter (formerly metadata)', () => {
    it('work as intended when substituded with an empty string', () => {
        const content = dedent`
            ---
            verbatim: {{ var }}
            quotedSingle: '{{ var }}'
            quotedDouble: "{{ var }}"
            ---

            # Some content.
        `;

        const liquidProcessed = liquidDocument(content, {var: ''}, 'frontmatter.test.ts.md');

        const [frontMatter] = extractFrontMatter(liquidProcessed);

        expect(frontMatter).toEqual({
            verbatim: '',
            quotedSingle: '',
            quotedDouble: '',
        });
    });

    it.each([
        {
            description: 'single quotes',
            content: dedent`
            ---
            quotes: '{{ withQuotes }}'
            ---
            Content
            `,
            vars: {withQuotes: "This isn't your typical substitution. It has single quotes."},
        },
        {
            description: 'double quotes',
            content: dedent`
            ---
            quotes: "{{ withQuotes }}"
            ---
            Content
            `,
            vars: {
                withQuotes: `"When you arise in the morning, think of what a precious privilege it is to be alive - to breathe, to think, to enjoy, to love." â€” Marcus Aurelius (allegedly)`,
            },
        },
        {
            description: 'curly braces',
            content: dedent`
            ---
            braces: {{ braces }}
            ---
            Content
            `,
            vars: {braces: '{}'},
        },
        {
            description: 'square brackets',
            content: dedent`
            ---
            brackets: {{ brackets }}
            ---
            Content
            `,
            vars: {brackets: '[]'},
        },
        {
            description: 'YAML multiline syntax',
            content: dedent`
            ---
            multiline: {{ multiline }}
            ---
            Content
            `,
            vars: {multiline: '>- This should break, right?\n\tRight?'},
        },
    ])(
        'should not fail even when variables contain reserved characters ($description)',
        ({content, vars}) => {
            const liquidProcessed = liquidDocument(content, vars, 'frontmatter.test.ts.md');

            expect(liquidProcessed).toMatchSnapshot();
        },
    );
});
