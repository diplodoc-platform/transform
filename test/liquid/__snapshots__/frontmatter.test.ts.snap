// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Liquid substitutions in front matter (formerly metadata) should not fail even when variables contain reserved characters (YAML multiline syntax) 1`] = `
"---
multiline: \\">- This should break, right?\\\\n\\\\tRight?\\"
---
Content"
`;

exports[`Liquid substitutions in front matter (formerly metadata) should not fail even when variables contain reserved characters (curly braces) 1`] = `
"---
braces: '{}'
---
Content"
`;

exports[`Liquid substitutions in front matter (formerly metadata) should not fail even when variables contain reserved characters (double quotes) 1`] = `
"---
quotes: '\\"When you arise in the morning, think of what a precious privilege it is to be alive - to breathe, to think, to enjoy, to love.\\" â€” Marcus Aurelius (allegedly)'
---
Content"
`;

exports[`Liquid substitutions in front matter (formerly metadata) should not fail even when variables contain reserved characters (single quotes) 1`] = `
"---
quotes: This isn't your typical substitution. It has single quotes.
---
Content"
`;

exports[`Liquid substitutions in front matter (formerly metadata) should not fail even when variables contain reserved characters (square brackets) 1`] = `
"---
brackets: '[]'
---
Content"
`;
